### PR TITLE
CI: Pin Cython on CI

### PR DESCRIPTION
**Pull request by phofl**
- [ ] ref #54756 (Replace xxxx with the GitHub issue number)
- [ ] [Tests added and passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#writing-tests) if fixing a bug or adding a new feature
- [ ] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).
- [ ] Added [type annotations](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#type-hints) to new arguments/methods/functions.
- [ ] Added an entry in the latest `doc/source/whatsnew/vX.X.X.rst` file if fixing a bug or adding a new feature.

### CODE REVIEW

**Review by MarcoGorelli — APPROVED** [review]

**Review by WillAyd — APPROVED** [review]

**Comment by WillAyd** [comment]
Thanks @phofl

**Comment by lithomas1** [comment]
Looks like numpy dev needs to be pinned as well, PR incoming.

File: ci/deps/actions-311-numpydev.yaml [hunk 1]
```diff
@@ -25,7 +25,7 @@ dependencies:
   - pip
 
   - pip:
-    - "cython"
+    - "cython<=3.0.1"
```
Comment (mroeschke) [inline ci/deps/actions-311-numpydev.yaml:28]: Don't we _not_ want 3.0.1?

**Comment by mroeschke** [comment]
I think we also need to pin cython in `.github/workflows/unit-tests.yml`

**Pull request merged**